### PR TITLE
rgw: make lua scripting optional

### PR DESCRIPTION
**Pull request by m-ildefons**
- introduce `WITH_RADOSGW_LUA` (default `ON`) compile time flag,
  toggling whether to build with Lua scripting support or not
- change default value of `WITH_RADOSGW_LUA_PACKAGES` from `ON` to the
  value of `WITH_RADOSGW_LUA`
- make Lua scripting support optional, depending on the value of
  `WITH_RADOSGW_LUA`

fixes: #55640
Signed-off-by: Moritz Röhrich <moritz.rohrich@suse.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/master/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by yuvalif** [comment]
not sure how we can handle: https://github.com/ceph/ceph/blob/master/src/test/cli/radosgw-admin/help.t
but the unit tests will probably fail if not matched to the changes

File: src/rgw/rgw_admin.cc [hunk 1]
```diff
@@ -290,12 +292,21 @@ void usage()
...
   cout << "  subscription pull          show events in a pubsub subscription\n";
   cout << "  subscription ack           ack (remove) an events in a pubsub subscription\n";
+#ifdef WITH_RADOSGW_LUA
   cout << "  script put                 upload a lua script to a context\n";
   cout << "  script get                 get the lua script of a context\n";
   cout << "  script rm                  remove the lua scripts of a context\n";
   cout << "  script-package add         add a lua package to the scripts allowlist\n";
   cout << "  script-package rm          remove a lua package from the scripts allowlist\n";
   cout << "  script-package list        get the lua packages allowlist\n";
+#else  // WITH_RADOSGW_LUA
```
[the remark this answers is not in the document]
Comment (m-ildefons) [reply src/rgw/rgw_admin.cc:302]: This is intentional and is supposed to keep discrepancies between documentation and the tool's help page small.
The average user of these command line tools is probably some sysadmin or operations person, not the one building the binaries. Therefore indicating at this level that the feature has been disabled avoids confusion about things that are in the documentation but don't end up being supported by the binaries in the end.

File: CMakeLists.txt [hunk 2]
```diff
@@ -418,7 +418,8 @@ option(WITH_RADOSGW "Rados Gateway is enabled" ON)
 option(WITH_RADOSGW_BEAST_OPENSSL "Rados Gateway's Beast frontend uses OpenSSL" ON)
 option(WITH_RADOSGW_AMQP_ENDPOINT "Rados Gateway's pubsub support for AMQP push endpoint" ON)
 option(WITH_RADOSGW_KAFKA_ENDPOINT "Rados Gateway's pubsub support for Kafka push endpoint" ON)
-option(WITH_RADOSGW_LUA_PACKAGES "Rados Gateway's support for dynamically adding lua packagess" ON)
+option(WITH_RADOSGW_LUA "Rados Gateway's support for Lua" ON)
+option(WITH_RADOSGW_LUA_PACKAGES "Rados Gateway's support for dynamically adding lua packagess" ${WITH_RADOSGW_LUA})
```
Comment (cbodley) [inline CMakeLists.txt:422]: i think `cmake_dependent_option()` is what you want for this one, see https://cmake.org/cmake/help/latest/module/CMakeDependentOption.html

Comment (m-ildefons) [reply CMakeLists.txt]: Thanks for pointing that out. I was not aware this existed, but it is indeed what I want.

**Comment by m-ildefons** [comment]
The tests should not fail, since Lua support has to be disabled by choice and it's enabled for the tests. I think the problem was the wrong usage of `option()` for `WITH_RADOSGW_LUA_PACKAGES` in the Cmake file, but if that's not it, I'll make the `--help` output static again and have the commands print out a useful message once invoked.

**Comment by yuvalif** [comment]
i think you should disable the lua unit test explicitly. 
for the case that someone build with lua disabled, but still want to run the other unit tests

**Comment by m-ildefons** [comment]
jenkins test windows

File: CMakeLists.txt [hunk 3]
```diff
@@ -418,7 +418,9 @@ option(WITH_RADOSGW "Rados Gateway is enabled" ON)
 option(WITH_RADOSGW_BEAST_OPENSSL "Rados Gateway's Beast frontend uses OpenSSL" ON)
 option(WITH_RADOSGW_AMQP_ENDPOINT "Rados Gateway's pubsub support for AMQP push endpoint" ON)
 option(WITH_RADOSGW_KAFKA_ENDPOINT "Rados Gateway's pubsub support for Kafka push endpoint" ON)
-option(WITH_RADOSGW_LUA_PACKAGES "Rados Gateway's support for dynamically adding lua packagess" ON)
+option(WITH_RADOSGW_LUA "Rados Gateway's support for Lua" ON)
+cmake_dependent_option(WITH_RADOSGW_LUA_PACKAGES
+  "Rados Gateway's support for dynamically adding lua packagess" ON "${WITH_RADOSGW_LUA}" OFF)
```
Comment (cbodley) [inline CMakeLists.txt:423]: it looks like this should be `"WITH_RADOSGW_LUA"`

File: src/CMakeLists.txt [hunk 4]
```diff
@@ -552,7 +552,7 @@ endif(${WITH_LTTNG})
 add_subdirectory(global)
 
 # RGW also support Lua scripting
-if(WITH_CEPHFS OR WITH_RADOSGW)
+if((WITH_CEPHFS OR WITH_RADOSGW) AND WITH_RADOSGW_LUA)
```
Comment (cbodley) [inline src/CMakeLists.txt:555]: i think the parens are wrong here. if cephfs requires lua, that shouldn't depend on WITH_RADOSGW_LUA:
```cmake
if(WITH_CEPHFS OR (WITH_RADOSGW AND WITH_RADOSGW_LUA))
```

**Comment by yuvalif** [comment]
@m-ildefons there are more lua context being added. see: rgw_lua_background.cc, rgw_ops.cc, rgw_main.cc
and there are more that will be added in the future. trying to ifdef them in all places that calls that will be difficult, and would break as they are added.
* for the purpose of security, blocking lua in `radosgw-admin` is sufficient
* for minimizing compile/runtime dependencies we should probably refactor the files so that the headers will just contain the "entry point" (e.g. the `execute` function). and stub these functions under the `ifdef` flag

**Pull request closed without merging**